pkg/controller/utils/utils.go

```diff
@@ -986,3 +986,53 @@ func RemoveInstallationFinalizer(i *operatorv1.Installation, finalizer string) {
 		i.SetFinalizers(stringsutil.RemoveStringInSlice(finalizer, i.GetFinalizers()))
 	}
 }
+
+// MaintainInstallationFinalizer manages a controller's finalizer on the Installation resource.
+// We add a finalizer to the Installation when the mainResource has been installed, and only remove that finalizer when
+// the resource has been deleted and its secondary resources have stopped running. This allows for a graceful cleanup of any resources
+// prior to the CNI plugin being removed.
+func MaintainInstallationFinalizer(
+	ctx context.Context,
+	c client.Client,
+	mainResource client.Object,
+	finalizer string,
+	secondaryResources ...client.Object,
+) error {
+	// Get the Installation.
+	installation := &operatorv1.Installation{}
+	if err := c.Get(ctx, DefaultInstanceKey, installation); err != nil {
+		if errors.IsNotFound(err) {
+			log.V(1).Info("Installation config not found")
+			return nil
+		}
+		log.Error(err, "An error occurred when querying the Installation resource")
+		return err
+	}
+	patchFrom := client.MergeFrom(installation.DeepCopy())
+
+	// Determine the correct finalizers to apply to the Installation.
+	if mainResource != nil {
+		// Add a finalizer indicating that the mainResource is still available.
+		SetInstallationFinalizer(installation, finalizer)
+	} else {
+		// Check if the namespaced secondaryResources are still present.
+		// Keep track of all the secondary resources that the main resource creates.
+		// Only delete the finalizer if all of the secondary resources are deleted.
+		for _, secondaryResource := range secondaryResources {
+			err := c.Get(ctx, types.NamespacedName{Namespace: secondaryResource.GetNamespace(), Name: secondaryResource.GetName()}, secondaryResource)
+			if err != nil && !errors.IsNotFound(err) {
+				return err
+			} else if errors.IsNotFound(err) {
+				log.Info(fmt.Sprintf("Object %s in Namespace %s no longer exists", secondaryResource.GetName(), secondaryResource.GetNamespace()), "finalizer", finalizer)
+			} else {
+				log.Info(fmt.Sprintf("Object %s in Namespace %s is still present, waiting for termination", secondaryResource.GetName(), secondaryResource.GetNamespace()))
```

Comment:
Not every object is going to have a namespace - throughout this function, the logging is going to be awkward if the namespace is an empty string.

We should handle that case explicitly - e.g.

```go
name := types.NamespacedName{Namespace: secondaryResource.GetNamespace(), Name: secondaryResource.GetName()}
```

Then you can use it in the logs:

```go
log.Info("%s is still present, waiting for termination", name)
```